Document updated anchor calculations in .NET 8.0 (#7985)
dotnet/winforms: docs/AnchorLayoutChangesInNet80.md

     {
        return;
     }

    if (!control.IsHandleCreated || !control.Parent.IsHandleCreated)
     {
        return;
     }

    ComputeAnchorInfo(IArrangedElement element)
}
```
## Simplifying Anchor calculations:
The current anchor calculation implementation is complicated and appears to have been a result of unnecessary attempts to calculate anchors with invalid bounds. Replacing the current implementation with the one described in Figure 1 above. Following is the source snippet that computes the anchors.

```CS
private static void ComputeAnchorInfo(IArrangedElement element)
 {
    AnchorInfo? anchorInfo = GetAnchorInfo(element);
    if (anchorInfo is null)
     {
        anchorInfo = new();
        SetAnchorInfo(element, anchorInfo);
     }

    Rectangle displayRect = element.Container.DisplayRectangle;
    Rectangle elementBounds = element.Bounds;

    int x = elementBounds.X;
    int y = elementBounds.Y;

    anchorInfo.Left = x;
    anchorInfo.Top = y;

    anchorInfo.Right = displayRect.Width - (x + elementBounds.Width);
    anchorInfo.Bottom = displayRect.Height - (y + elementBounds.Height);
 }
```
## Risk mitigation:

> Layout in general is complex and could impact every component in the WinForms  . In order to reduce the potential risk and provide backward compatibility, This changes are `quirked` under switch `System.Windows.Forms.EnableAnchorLayoutV2`. These changes are by default on for new/migrating applications targeting .NET 8.0 but the developers can control this by setting the above mentioned flag to `false` in the runtimeconfig .template.json for the application.

Reviewer: I got a similar feedback in the blog draft - we don't appear to be using "quirk" publicly.

Author: how about `hidden` under flag?

# Anchor layout changes in .NET 8.0
## Description:

> We have multiple [issues](https://github.com/dotnet/winforms/issues?q=is%3Aissue+is%3Aopen+anchor+-label%3Atest-bug+-label%3A%22%3Aconstruction%3A+work+in+progress%22+-label%3Aapi-suggestion+-label%3Aapi-approved+-label%3Atenet-localization) reported in WinForms around the anchor layout being problematic on higher DPI scale monitors, irrespective of application DPI mode. This document outlines the changes being made in .NET 8.0 to address these issues along with setting the goal to support all supported application DPI modes in WinForms.

Author: thx for pointing this. I remember using anchor label in query.

## Problem in Scope:

> Anchored control’s position with respect to its parent should be able to determine at the design time and would only need to be changed   if there were explicit changes in the control’s Bounds or when the control is scaled in response to a DPI changed event. Bounds changes as result of Parent’s bounds change shouldn’t alter control’s relative position in the parent’s rectangle. However, layout in WinForms computes the anchored control’s position every time there are changes to control’s bounds or control’s property change that may impact its position. This is leading to the [issues](https://github.com/dotnet/winforms/issues?q=is%3Aissue+is%3Aopen+anchor+-label%3Atest-bug+-label%3A%22%3Aconstruction%3A+work+in+progress%22+-label%3Aapi-suggestion+-label%3Aapi-approved+-label%3Atenet-localization)  we have been seeing . The following is a source snippet that is being serialized in WinForms designer and added comments show the various events that trigger anchor computations and why they could be wrong or unnecessary.

Reviewer: The 3rd and onwards:
```suggestion
Anchored control’s position with respect to its parent should be able to determine at the design time and would only need to be changed   if there were explicit changes in the control’s Bounds or when the control is scaled in response to a DPI changed event. Bounds changes as result of Parent’s bounds change shouldn’t alter control’s relative position in the parent’s rectangle. However, the layout engine computes the anchored control’s position every time there are changes to the control’s bounds or the control's location-related properties.

For example, the following somewhat trivial code snippet copide from an `InitializeComponent` method demonstrates the impact of the "over-eager" computations. The number of unncessary computations can increase quite dramatically for nested layouts.
```

```CS
// Following line trigger layout to compute anchors with default button size and without parent.
this.button7.Anchor = ((System.Windows.Forms.AnchorStyles)((System.Windows.Forms.AnchorStyles.Left | System.Windows.Forms.AnchorStyles.Right)));

// Following line trigger layout again to compute anchors with default button size with new location and still without parent.
this.button7.Location = new System.Drawing.Point(9, 47);

// Following line trigger layout again to compute anchors with button's new size and still without parent.
this.button7.Size = new System.Drawing.Size(134, 23);

// Following line trigger layout to compute anchors with button's current size and default size for parent.
this.Controls.Add(button7);

// Following line trigger layout to compute anchors with button's current size and new size for parent.
// These size still may be changed for the DPI on the monitor depending on application's DPI mode
this.Size = new System.Drawing.Size(828, 146);

// Following line trigger layout to compute anchors with button's
current size and new size for parent.
// DPI on the monitor still not applied.
this.ResumeLayout(false)
```

> The above snippet does not represent the complete set of instances where anchor computations are unnecessary and may hold invalid anchor values. It gets even more complicated when nested UserControls are involved.

Reviewer: I would remove this sentence.

Author: why?

Reviewer 2: FYI - I do like this sentence and agree that keeping it is a good idea.

## Known issues:

> We have multiple issues reported [here](https://github.com/dotnet/winforms/issues?q=is%3Aissue+is%3Aopen+anchor+-label%3Atest-bug+-label%3A%22%3Aconstruction%3A+work+in+progress%22+-label%3Aapi-suggestion+-label%3Aapi-approved+-label%3Atenet-localization) from customers and some of them are direct result of anchor miscalculations. The following are snippets of the category of issues we currently see.

Reviewer: We seem to be repeating this idea over and over. 
```suggestion
Customer reported issues can be largely groupped as follows.
```

### Missing controls:
Calculation of control's anchors with default sizes may result in negative anchors and thus result in invalid location for the anchored controls.

![MissingControls](images/AnchorLayoutKnownIssue_MissingControl.png)
### Overlapped Controls:

> Along with above reason, If we happened to scale parent control to match with current monitors DPI and child control's anchor calculations happen out of sync with this scaling, we may see issues like in below pic.

Reviewer: Refrain from "we", "you", etc., and use the passive voice.
```suggestion
Related to the above reason, if the parent control is caled to match the current monitor's DPI, and the control's anchor calculations happen out of sync with this scaling, controls may overlap:
```

![OverlappedControls](images/AnchorLayoutKnownIssue_OverlappedControl.png)
## Anchor Calculations:

> The following image illustrates how we calculate anchors with respect to a control’s parent’s display rectangle. Rectangle marked in `red` is Parent’s display rectangle while the rectangle marked in `green` is anchored control’s (button) bounds.

Reviewer: ```suggestion
The following image illustrates anchors calculation with respect to a control’s parent’s display rectangle. The "red" rectangle is the parent’s display rectangle, and the "green" rectangle is the anchored control’s (button) bounds.
```

Ditto below

- `Left` arrow indicates the X co-ordinate of the button location with respect to parents display rectangle.

> - `Top` arrow indicates the Y co-ordinate of the button location with respect to parents display rectangle.

Reviewer: "coordinate" doesn't have hyphers.

- `Right` arrow indicates the distance from right edge of parent’s rectangle where button is placed.
- `Right` arrow indicates the distance from bottom edge of parent’s rectangle where button is placed.

    ![AnchorCalculations](images/AnchorCalculations.png)

> In the above picture the boundary marked in `red` is a container hosting the control (boundary marked with `green`). The control can define its anchor property to explicitly tell the layout engine how its bounds change relative to its parent’s bounds. Ex. 

Reviewer: ```suggestion
In the above picture the boundary marked in `red` is a container hosting the control (boundary marked with `green`). The control can define its anchor property to explicitly tell the layout engine how its bounds change relative to its parent’s bounds. For example:
```

```CS
this.button14.Anchor = (System.Windows.Forms.AnchorStyles.Bottom
 | System.Windows.Forms.AnchorStyles.Left
 | System.Windows.Forms.AnchorStyles.Top
 | System.Windows.Forms.AnchorStyles.Right );
```

> When Control’s Anchor property is set, we compute anchors (left, top, right, bottom values) and store them in an internal struct `AnchorInfo`. We expect these anchor values to be in negative only when the button is placed/position outside/overlapped with the hosting controls bounds.

Reviewer: ```suggestion
When the control’s `Anchor` property is set, the anchors (that is, left, top, right, bottom values) are computed and stored in an internal struct `AnchorInfo`. The only time the anchor values can be negative is when the control is placed/position outside/overlapped with the hosting controls bounds.
```

## Proposed solution:

> Unnecessary control’s anchors computation is the root cause for many [issues](https://github.com/dotnet/winforms/issues?q=is%3Aissue+is%3Aopen+anchor+-label%3Atest-bug+-label%3A%22%3Aconstruction%3A+work+in+progress%22+-label%3Aapi-suggestion+-label%3Aapi-approved+-label%3Atenet-localization) reported so far. In this proposal, we are delaying the anchors computation to the specific time in the layout flow - which is when the control’s and its parent’s handles are created. By that time, in the majority of scenarios, the control’s and its parent’s bounds are finalized and the display monitor’s DPI has been applied to them.  We also have issues with how we are calculating anchors but, in this proposal, we are simplifying anchor calculations as mentioned in the above pic to help diagnose any future issues.

Reviewer: IMO the first sentence is redundant, the previous sections set the scene and described the issue. This section is expected to describe the solution.

```suggestion
The proposal is to delay the anchors computation until the control’s and its parent’s handles are created. By that time, in the majority of scenarios, the control’s and its parent’s bounds are finalized, and the display monitor’s DPI has been applied to them.  Whilst there may still be other issues related to how anchors are calculated, but this proposal would singnificantly improve our ability to diagnose and address such issues.
```

> We may still have cases where developers are forced to create handle explicitly, out of order, but those cases can be handled separately   by the application developer for any anchor miscalculations. The following are the events we would be using to compute anchors and replacing the current set of events mentioned in Scope section above.
> - OnCreateControl
> - OnParentChanged
> - SetBounds

Reviewer: ```suggestion
Additionally there may be cases where developers could be force-creating handle out of order, but those cases are not expected to be main stream, and such cases will have to be handled by application developer. 

The following are the events we would be using to compute anchors and replacing the current set of events mentioned in Scope section above.
- `OnCreateControl`,
- `OnParentChanged`,
- `SetBounds`.
```

### Source snippet:

```CS
protected virtual void OnCreateControl()
{
    DefaultLayout.UpdateAnchorInfoV2(this);
}

internal static void UpdateAnchorInfoV2(IArrangedElement element)
{
    if (!LocalAppContextSwitches.EnableAnchorLayoutV2
     || !CommonProperties.GetNeedsAnchorLayout(element))
     {
        return;
     }

    Control control = element as Control;
    Debug.Assert(control != null, "AnchorLayoutV2 and beyond are expected to be used only on Control type");

    if (control is null || control.Parent is null)
     {
        return;
     }

    if (!control.IsHandleCreated || !control.Parent.IsHandleCreated)
     {
        return;
     }

    ComputeAnchorInfo(IArrangedElement element)
}
```
## Simplifying Anchor calculations:

> The current anchor calculation implementation is complicated and appears to have been a result of unnecessary attempts to calculate anchors with invalid bounds. Replacing the current implementation with the one described in Figure 1 above. Following is the source snippet that computes the anchors.

Reviewer: The first sentense feels like we're airing dirty laundy, I suggest removing it.

```CS
private static void ComputeAnchorInfo(IArrangedElement element)
 {
    AnchorInfo? anchorInfo = GetAnchorInfo(element);
    if (anchorInfo is null)
     {
        anchorInfo = new();
        SetAnchorInfo(element, anchorInfo);
     }

    Rectangle displayRect = element.Container.DisplayRectangle;
    Rectangle elementBounds = element.Bounds;

    int x = elementBounds.X;
    int y = elementBounds.Y;

    anchorInfo.Left = x;
    anchorInfo.Top = y;

    anchorInfo.Right = displayRect.Width - (x + elementBounds.Width);
    anchorInfo.Bottom = displayRect.Height - (y + elementBounds.Height);
 }
```
## Risk mitigation:

> Layout in general is complex and could impact every component in the WinForms. In order to reduce the potential risk and provide backward compatibility, These changes are `hidden` under switch `System.Windows.Forms.EnableAnchorLayoutV2` and are `ON` by default for new/migrating applications targeting .NET 8.0. Developers can disable these changes by setting the above mentioned flag to `false` in the runtimeconfig.template.json for the application.

Reviewer: "Layout ...is complex and could impact every component in the WinForms." - Yes, the layout is complex and impacts every control. What's the point? :)

```suggestion
The layout engine, in general, is quite complex, and any changes in it carry a very high risk. In order to reduce the risks and to provide backward compatibility, the new anchor calculations are put behind a feature switch `System.Windows.Forms.EnableAnchorLayoutV2`. The switch is enabled by default for Windows Forms applications targeting .NET 8.
It is possible to retain the original anchor calculations by setting the swtich to `false` in the [runtimeconfig.json](https://learn.microsoft.com/dotnet/core/runtime-config/#runtimeconfigjson).
```

Author: Setting justification for introducing switch.

Author: > "Layout ...is complex and could impact every component in the WinForms." - Yes, the layout is complex and impacts every control. What's the point? :)

Setting justification for introducing switch.

> # Anchor layout changes in .NET 8.0

Reviewer: Reading https://github.com/dotnet/winforms/pull/7956 I realised that the proposed location is probably not the best. The docs folder contains general purpose docs, e.g., related to how to build the repo. This doc is a "feature design" doc, and I don't think it belongs here. Especially, as other designs docs will be added it will create a mental overhead for both maitnainers and contributors.
For example, have a look at how the docs are structured in https://github.com/dotnet/runtime/tree/main/docs.

I suggest creating "docs/design" folder and moving this doc there.

Another consideration to make is that CamelCase could make it harder to read file names, as well as non-Windows OS are case sensitive when it comes to file names. With this, we adopted a "kebab-casing" for file names, and dotnet/runtime generally uses the same kebab-casing for naming docs too.

## Description

> We have multiple [issues](https://github.com/dotnet/winforms/issues?q=is%3Aissue+is%3Aopen+anchor+label%3A%22area%3A+anchor%2Fscaling%22) reported in WinForms around the anchor layout being problematic on higher DPI scale monitors, irrespective of application DPI mode. This document outlines the changes being made in .NET 8.0 to address these issues along with setting the goal to support all supported application DPI modes in WinForms.

Reviewer 2: ```suggestion
We have multiple [issues](https://github.com/dotnet/winforms/issues?q=is%3Aissue+is%3Aopen+anchor+label%3A%22area%3A+anchor%2Fscaling%22) reported in WinForms around the anchor layout being problematic on higher DPI scale monitors, irrespective of application DPI mode. This document outlines the changes being made in .NET 8.0 to address these issues, which is aligned with the larger goal to support all supported application DPI modes in WinForms.
```

## Problem in Scope

> Anchored control's position with respect to its parent should be able to determine at the design time and would only need to be changed   if there were explicit changes in the control's Bounds or when the control is scaled in response to a DPI changed event. Bounds changes as result of Parent's bounds change shouldn’t alter control's relative position in the parent's rectangle. However, the layout engine computes the anchored control's position every time there are changes to the control's bounds or the control's location-related properties.

Reviewer 2: ```suggestion
The position of an anchored control with respect to its parent should be able to determined at design time and should only need to be changed  if there were explicit changes in the control's bounds or when the control is scaled in response to a DPI changed event. Bounds changes as result of the parent control's bounds changing shouldn’t alter a control's relative position in the parent's rectangle. However, currently the layout engine computes the anchored control's position every time there are changes to the control's bounds or the control's location-related properties.
```

> The position of an anchored control with respect to its parent should be able to determined at design time and should only need to be changed  if there were explicit changes in the control's bounds or when the control is scaled in response to a DPI changed event. Bounds changes as result of the parent control's bounds changing shouldn’t alter a control's relative position in the parent's rectangle. However, currently the layout engine computes the anchored control's position every time there are changes to the control's bounds or the control's location-related properties.

Reviewer: Looks like there are rogue apostrophies, may worth bulk fixing this
![image](https://user-images.githubusercontent.com/4403806/197901940-35377b8c-058f-4884-b71b-b959148ad16e.png)

> For example, the following somewhat trivial code snippet copide from an `InitializeComponent` method demonstrates the impact of the "over-eager" computations. The number of unncessary computations can increase quite dramatically for nested layouts.

Reviewer 2: ```suggestion
For example, the following simple code snippet copied from an `InitializeComponent` method demonstrates the impact of the "over-eager" computations. The number of unnecessaryy computations can increase quite dramatically for nested layouts.
```

```CS

> // Following line trigger layout to compute anchors with default button size and without parent.

Reviewer 2: ```suggestion
// The following line triggers layout to compute anchors with default button size and without parent.
```

this.button7.Anchor = ((System.Windows.Forms.AnchorStyles)((System.Windows.Forms.AnchorStyles.Left | System.Windows.Forms.AnchorStyles.Right)));

// Following line trigger layout again to compute anchors with default button size with new location and still without parent.
this.button7.Location = new System.Drawing.Point(9, 47);

> // Following line trigger layout again to compute anchors with button's new size and still without parent.

Reviewer 2: ```suggestion
// The following line triggers layout once again to compute anchors with button's new size but still without parent.
```

this.button7.Size = new System.Drawing.Size(134, 23);

// Following line trigger layout to compute anchors with button's current size and default size for parent.
this.Controls.Add(button7);

> // Following line trigger layout to compute anchors with button's current size and new size for parent.

Reviewer 2: ```suggestion
// The following line triggers layout to compute anchors with button's current size and new size for parent.
```

> // These size still may be changed for the DPI on the monitor depending on application's DPI mode

Reviewer 2: ```suggestion
// These sizes still may be changed for the DPI on the monitor depending on application's DPI mode
```

this.Size = new System.Drawing.Size(828, 146);

> // Following line trigger layout to compute anchors with button's

Reviewer 2: ```suggestion
// The following line triggers layout to compute anchors with button's
```

current size and new size for parent.
// DPI on the monitor still not applied.
this.ResumeLayout(false)
```
The above snippet does not represent the complete set of instances where anchor computations are unnecessary and may hold invalid anchor values. It gets even more complicated when nested UserControls are involved.

## Known issues

Customer reported issues can be largely groupped as follows.

### Missing controls

Calculation of control's anchors with default sizes may result in negative anchors and, thus, result in invalid location for the anchored controls.

![MissingControls](images/AnchorLayoutKnownIssue_MissingControl.png)

### Overlapped Controls

> Related to the above reason, if the parent control is caled to match the current monitor's DPI, and the control's anchor calculations happen out of sync with this scaling, controls may overlap:

Reviewer: ```suggestion
Related to the above reason, if the parent control is scaled to match the current monitor's DPI, and the control's anchor calculations happen out of sync with this scaling, controls may overlap:
```

![OverlappedControls](images/AnchorLayoutKnownIssue_OverlappedControl.png)

## Anchor Calculations

> The following image illustrates anchors calculation with respect to a control's parent's display rectangle. The "red" rectangle is the parent's display rectangle, and the "green" rectangle is the anchored control's (button) bounds.

Reviewer 3: Would you be able to replace green/red color combination with some other colors or pen patterns?

Reviewer: Great a11y reminder! Thank you

- `Left` arrow indicates the X coordinate of the button location with respect to parents display rectangle.
- `Top` arrow indicates the Y coordinate of the button location with respect to parents display rectangle.
- `Right` arrow indicates the distance from right edge of parent's rectangle where button is placed.
- `Right` arrow indicates the distance from bottom edge of parent's rectangle where button is placed.

    ![AnchorCalculations](images/AnchorCalculations.png)

> In the above picture the boundary marked in `red` is a container hosting the control (boundary marked with `green`). The control can define its anchor property to explicitly tell the layout engine how its bounds change relative to its parent's bounds. For example: 

Reviewer: ```suggestion
In the above image the boundary marked in "red" is a container hosting the control (boundary marked with "green"). The control can define its anchor property to explicitly tell the layout engine how its bounds change relative to its parent's bounds. For example: 
```

```CS
this.button14.Anchor = (System.Windows.Forms.AnchorStyles.Bottom
 | System.Windows.Forms.AnchorStyles.Left
 | System.Windows.Forms.AnchorStyles.Top
 | System.Windows.Forms.AnchorStyles.Right );
```
When the control's `Anchor` property is set, the anchors (that is, left, top, right, bottom values) are computed and stored in an internal struct `AnchorInfo`. The only time the anchor values can be negative is when the control is placed/position outside/overlapped with the hosting controls bounds.

## Proposed solution

> Additionally there may be cases where developers could be force-creating handle out of order, but those cases are not expected to be main stream, and such cases will have to be handled by application developer. 

Reviewer 2: ```suggestion
There may be cases where developers could be force-creating handles out of order, but those cases are not expected to be mainstream, and such cases will have to be handled by application developer. 
```

> The following are the events we would be using to compute anchors and replacing the current set of events mentioned in Scope section above.

Reviewer 2: ```suggestion
The following are the events we will be using to compute anchors and replacing the current set of events mentioned in Scope section above.
```

- `WmCreate`,
- `OnParentChanged`,
- `SetBounds`.

### Source snippet

```CS
private void WmCreate()
{
     ...
     DefaultLayout.UpdateAnchorInfoV2(this);
}

internal static void UpdateAnchorInfoV2(IArrangedElement element)
{
    if (!LocalAppContextSwitches.EnableAnchorLayoutV2
     || !CommonProperties.GetNeedsAnchorLayout(element))
     {
        return;
     }

    Control control = element as Control;
    Debug.Assert(control != null, "AnchorLayoutV2 and beyond are expected to be used only on Control type");

    if (control is null || control.Parent is null)
     {
        return;
     }

    if (!control.IsHandleCreated || !control.Parent.IsHandleCreated)
     {
        return;

>      }

Reviewer: The indentations here and in other snippets appear off.

    ComputeAnchorInfo(IArrangedElement element)
}
```

## Simplifying Anchor calculations

> The current anchor calculation implementation is designed prior to adding high DPI support to winforms. This proposal is replacing the current implementation with the one described in Figure 1 above. Following is the source snippet that computes the anchors.

Reviewer 2: ```suggestion
The current anchor calculation implementation was designed prior to adding high DPI support to WinForms. This proposal is replacing the current implementation with the one described in Figure 1 above. Following is the source snippet that computes the anchors.
```

```CS
private static void ComputeAnchorInfo(IArrangedElement element)
 {
    AnchorInfo? anchorInfo = GetAnchorInfo(element);
    if (anchorInfo is null)
     {
        anchorInfo = new();
        SetAnchorInfo(element, anchorInfo);

>      }

Reviewer: ditto

    Rectangle displayRect = element.Container.DisplayRectangle;
    Rectangle elementBounds = element.Bounds;

    int x = elementBounds.X;
    int y = elementBounds.Y;

    anchorInfo.Left = x;
    anchorInfo.Top = y;

    anchorInfo.Right = displayRect.Width - (x + elementBounds.Width);
    anchorInfo.Bottom = displayRect.Height - (y + elementBounds.Height);
 }
```

## Risk mitigation

The layout engine, in general, is quite complex, and any changes in it carry a very high risk. In order to reduce the risks and to provide backward compatibility, the new anchor calculations are put behind a feature switch `System.Windows.Forms.EnableAnchorLayoutV2`. The switch is enabled by default for Windows Forms applications targeting .NET 8.

It is possible to retain the original anchor calculations by setting the swtich to `false` in the [runtimeconfig.json](https://learn.microsoft.com/dotnet/core/runtime-config/#runtimeconfigjson).
Snippet for runtimeconfig.template.json:
```JSON
{
  "configProperties": {

>     "System.Windows.Forms.EnableAnchorLayoutV2": false

Reviewer: ❗  Please ensure the switch name is aligned with the actual implementation (#7956)
```suggestion
    "System.Windows.Forms.AnchorLayoutV2": false
```
